tests/integration/security/authorization_test.go

```diff
@@ -1212,28 +1213,80 @@ func TestAuthorization_Audit(t *testing.T) {
 					ExpectAllowed: expectAllowed,
 				}
 			}
-			cases := []rbacUtil.TestCase{
-				newTestCase(b, "/allow", true),
-				newTestCase(b, "/audit", false),
-				newTestCase(c, "/audit", true),
-				newTestCase(c, "/deny", false),
-				newTestCase(d, "/audit", true),
-				newTestCase(d, "/other", true),
-			}
 
-			args := map[string]string{
-				"Namespace":     ns.Name(),
-				"RootNamespace": istio.GetOrFail(t, t).Settings().SystemNamespace,
-			}
+			for _, a := range []echo.Instances{apps.A.Match(echo.Namespace(ns.Name())), apps.VM.Match(echo.Namespace(ns.Name()))} {
+				cases := []rbacUtil.TestCase{
+					newTestCase(a, b, "/allow", true),
+					newTestCase(a, b, "/audit", false),
+					newTestCase(a, c, "/audit", true),
+					newTestCase(a, c, "/deny", false),
+					newTestCase(a, d, "/audit", true),
+					newTestCase(a, d, "/other", true),
+				}
+				t.NewSubTest(fmt.Sprintf("from %s in %s", a[0].Config().Service, a[0].Config().Cluster.StableName())).
+					Run(func(t framework.TestContext) {
+						args := map[string]string{
+							"b":             b[0].Config().Service,
+							"c":             c[0].Config().Service,
+							"d":             d[0].Config().Service,
+							"Namespace":     ns.Name(),
+							"RootNamespace": istio.GetOrFail(t, t).Settings().SystemNamespace,
+						}
+						applyPolicy := func(filename string, ns namespace.Instance) {
+							policy := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, filename))
+							t.Config().ApplyYAMLOrFail(t, ns.Name(), policy...)
+						}
+						applyPolicy("testdata/authz/v1beta1-audit.yaml.tmpl", ns)
 
-			applyPolicy := func(filename string, ns namespace.Instance) {
-				policy := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, filename))
-				t.Config().ApplyYAMLOrFail(t, ns.Name(), policy...)
+						rbacUtil.RunRBACTest(t, cases)
+					})
 			}
 
-			applyPolicy("testdata/authz/v1beta1-audit.yaml.tmpl", ns)
+			// (TODO)JimmyCYJ: Support multiple VMs and apply audit policies to multiple VMs for testing.
```

Comment:
can you also mention the following duplicate is really a short hack and should be removed (merged) with the above test cases?

Reply:
Done.